### ISSUE TITLE
Expose htmlparser2 (or allow a custom processNodes function)

### ISSUE THREAD

**Issue by daviestar** [wrakky/react-html-parser#26]
Hey,

For what I'm trying to achieve, I need to take multiple passes at HTML parsing.  Since the `transform` function of this lib returns react elements, I need to bring in the `htmlparser2` lib as a separate dependency so I can do this, then use `ReactHtmlParser.processNodes`.

Cheers.

**EDIT**

Oops.. `ReactHtmlParser.processNodes` should have been `ReactHtmlParser.convertNodeToElement `

**Comment by wrakky** [comment]
Sorry I don't entirely follow what the question is, or what you are trying to do. Can you elaborate - maybe with a code example if you think that would help?

**Comment by daviestar** [comment]
I'll try to explain :) Essentially, I have a heap of HTML from the Wikipedia API which is mostly on one level like this:
```html
<h2>...</h2>
<h3>...</h3>
<p>...</p>
<p>...</p>
<p>...</p>
<h3></h3>
<h2></h2>
```
See that last `h3` has no content?  I'm trying to remove those titles.  There are more gnarly situations like this:
```html
<h2>...</h2>
<h3>...</h3>
<p>...</p>
<h2></h2>
<h3></h3>
<h3></h3>
<h3></h3>
<h2></h2>
<p>...</p>
```
First of all, I've found htmlparser2's `prev` and `next` node attributes are unreliable and sometimes have null values when they shouldn't, when I try a recursive pattern like node > next > next > next > next...

 So I've been able to achieve my goal using a basic loop and side effects, but I need access to the array of htmlparser2 nodes, like:

```js
function processWiki(html) {
  const nodes = parseDOM(html, { decodeEntities: true })
  return processNodes(nodes).map(convertNodeToElement)
}

function processNodes(nodes, { showPruned }) {
  // ...
  for (const node of nodes) {
```

So basically - like the title suggests - it would be nice if your lib could just expose the htmlparser2 lib so I don't need to bring it into my project separately.  I'm currently using it just fine like this:

**Comment by wrakky** [comment]
So what you're saying is?
- `react-html-parser` has a dependency on `htmparser2`
- You need to do some custom stuff that requires `htmlparser2` directly but don't want to add it as a dependency to your project when it's already a dependency of one of your other dependencies

Is that right?

**Comment by daviestar** [comment]
Sorry I've updated my reply about a bazillion times after too much coffee, so I'm not sure which version you've actually read :)

Yes, that would be great - but perhaps a better solution would be the ability to have a custom `processNodes` function, like my last edit.

**Comment by wrakky** [comment]
OK I think allowing the `processNodes` function to be overridden is the best solution as that gives you complete control but with the caveat that it's for advanced use and you have to join everything up yourself.

So you can do something like this to totally ignore the internal `processNodes` and do everything yourself:
```js
import ReactHtmlParser, { convertNodeToElement } from 'react-html-parser'
const elements = ReactHtmlParser(html, { processNodes })
function processNodes(nodes) {
  return nodes
    .customProcessing() // return array of nodes
    .map((node, index) => convertNodeToElement(node, index));
}
```

Or you can even join up with the internal `processNodes` after your custom processing which will allow you to process and filter the nodes as you want but still keep the filtering, transforming and mapping that the internal `processNodes` gives you:
```js
import ReactHtmlParser, { processNodes } from 'react-html-parser'
const elements = ReactHtmlParser(html, { processNodes: myProcessNodes })
function myProcessNodes(nodes, transform) {
  const customNodes = customProcessing(nodes); // return array of nodes
  return processNodes(customNodes, transform);
}
```

**Comment by daviestar** [comment]
Yeah I think that would be a nice addition to your library 👍

**Comment by wrakky** [comment]
I'm going to have to consider how best to approach this. [TagElementType](https://github.com/wrakky/react-html-parser/blob/master/src/elementTypes/TagElementType.js) calls `processNodes` on the node children which adds complexity to the situation as to make it recursive will mean passing `processNodes` to itself so it can pass it on down the chain which I don't like the sound of.

I think a better approach now would be to add a `preprocessNodes` option which will be passed the initial nodes for preprocessing before delegating back to the library:

```js
export default function HtmlParser(html, { decodeEntities=true, transform, preprocessNodes }={}) {
  let nodes = htmlparser2.parseDOM(html, { decodeEntities });
  if (preprocessNodes) {
    nodes = preprocessNodes(nodes);
  }
  return processNodes(nodes, transform);
}
```

**Comment by daviestar** [comment]
Hey @wrakky that seems like a solution that would work, but I think it could be improved by refactoring `processNodes` to call itself in the function body - pulling in the if-logic from `TagElementType`.  Then in my case, it would be up to the user to write a custom recursive `processNodes` function.

Although this adds a bit of complexity for the user, to me it makes sense since we're dealing with tree structures, and would give the most flexibility.

**Comment by ayyazdaniaryan** [comment]
I'm joining in this conversation because I also agree that having access to a `processNodes` or even the internal `htmlparser2` would be supremely helpful. It would make traversing the nodes much easier. In my case, I've created an xml-like DSL and would like to traverse the elements, filter and replace nodes prior to finally attaching the resulting elements to the DOM via react. Having access to the htmlparser or even pre-processing nodes would certainly help.

**Comment by wrakky** [comment]
I've just released v2.0.1 with a `preprocessNodes` option and `htmlparser2` exposed. I think this is the best balance between flexibility and complexity and should allow more advanced use cases like the ones mentioned here.
